mantidproject/mantid
# Add ORSO Nexus save option to ISIS Reflectometry GUI

### Description of work

#### Summary of work

This PR extends both the `SaveISISReflectometryORSO` algorithm and the ISIS Reflectometry GUI to support saving into the ORSO Nexus format.



Fixes #37162. 




#### Further detail of work

- The orsopy library has a `save_nexus` method that takes the same data structure that you create when saving into the ASCII format but instead outputs it as a Nexus file. This PR updates the `SaveISISReflectometryORSO` algorithm to use that method to save into the Nexus format when a `.orb` extension has been given on the input filepath.

- The Save Ascii tab in the Reflectometry GUI has been extended to add ORSO Nexus as an additional file format option. The same set of checkbox options should be available when the ORSO Nexus format is selected as are available for the ORSO Ascii format. I have re-named the Save Ascii tab to just the Save tab to reflect that it now supports non-Ascii formats and have updated all user-facing references to Save Ascii in the software and documentation.

- Due to time constraints with code freeze, I haven't changed the name of the AsciiSaver class in the code yet, but will be opening a maintenance issue to do that.

### To test:


1) Go to `Interfaces` -> `Reflectometry` -> `ISIS Reflectometry`.
1) Go to the Save tab. Set a Save Path for saving the file to.
1) In the File Format section, select `ORSO Nexus (*.orb)` from the drop-down.
1) In the Automatic Save section, tick the `Save automatically on completion` checkbox.
1) The `Include individual row outputs for groups` checkbox should be enabled. Select this checkbox.
1) The `Save multiple datasets to a single file` checkbox should be enabled and selected by default.
1) Go to the Runs tab, enter Investigation ID `1120015` and Cycle `11_3`. Click the Search button.
1) From the search results, click to highlight runs `11934`, `11935` and `11936` and click the Transfer button to transfer them into the batch table on the right hand side. Click the Process button to reduce the data.
1) If you navigate to the save path that you set in the Save tab, you should see a single `.orb` file with the suffix `_multi` in the filename. When viewed in an HDF5 viewer, this file should contain multiple datasets. Assuming you have left the `Additional columns (includes Q resolution)` checkbox ticked, the non-stitched (individual angle) datasets should have 8 columns and the final stitched dataset should have 4 columns.
1) Test that the checkboxes on the Save tab are enabled and disabled correctly for the ORSO Nexus format (checkbox availability and behaviour should be the same as for the ORSO Ascii format - see manual testing documentation changes in this PR for further details, if required).





---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.
